File: BackUpForABook/UIControls/TRClasses/TRSliceGrid.m
//
//  TRSliceGrid.m
//  ATreeRingMap
//
//  Created by Hui Wang on 5/24/14.
//  Copyright (c) 2014 AnnieLyticx. All rights reserved.
//

#import "TRSliceGrid.h"

@interface TRSliceGrid (PrivateMethods)
// paint grid background
-(void)paintGridRingBackground:(CGContextRef)ctx
                        layout:(Slice)size
                        center:(CGPoint)origin
                      ringAxis:(TRAxis *)ringAxis;
// paint grid lines
-(void)paintGridLines:(CGContextRef)ctx
               layout:(Slice)size
               center:(CGPoint)origin
             ringAxis:(TRAxis *)ringAxis
            angleAxis:(TRAxis *)angleAxis;

// paint grid table cells
-(void)paintCells:(CGContextRef)ctx
            table:(TRDataTableSliceView*) table
           layout:(Slice)size
           center:(CGPoint)origin
         ringAxis:(TRAxis *)ringAxis
        angleAxis:(TRAxis *)angleAxis;

// paint grid cell at given location
-(void)paintCell:(CGContextRef)ctx
            cell:(TRCell *)cell
        position:(CGPoint)position;

// dynamicly draw grid cell at given location
-(void)drawCell:(CGContextRef)ctx
           cell:(TRCell *)cell
       position:(CGPoint)position;

// get cell path
-(UIBezierPath *)getCellPath:(TRCell *)cell
                    position:(CGPoint)position;

@end

@implementation TRSliceGrid

-(id)init
{
    self = [super init];
    
    _cellShowAsLayer = TRUE;
    
    // create grid line objects
    _columnGridLine = [[TRGridLine alloc] init];
    _rowGridLine = [[TRGridLine alloc] init];
    
    // initial row and column focus
    _rowFocus = -1;
    _columnFocus = -1;
    
    return self;
}

// paint grid and grid cell tabel on the given context
-(void)paint:(CGContextRef)ctx
       table:(TRDataTableSliceView*) table
      layout:(Slice)size
      center:(CGPoint)origin
    ringAxis:(TRAxis *)ringAxis
   angleAxis:(TRAxis *)angleAxis
{
    // (1) paint grid background
    CGContextSaveGState(ctx);
    [self paintGridRingBackground:ctx
                           layout:size
                           center:origin
                         ringAxis:ringAxis];
    CGContextRestoreGState(ctx);

    // (2) paint grid lines
    CGContextSaveGState(ctx);
    [self paintGridLines:ctx
                  layout:size
                  center:origin
                ringAxis:ringAxis
               angleAxis:angleAxis];
    CGContextRestoreGState(ctx);
    
    // (3) paint grid cells
    if (_cellShowAsLayer == FALSE)
    {
        CGContextSaveGState(ctx);
        [self paintCells:ctx
               table:table
              layout:size
              center:origin
            ringAxis:ringAxis
           angleAxis:angleAxis];
        CGContextRestoreGState(ctx);
    }
}

// paint grid background
-(void)paintGridRingBackground:(CGContextRef)ctx
                        layout:(Slice)size
                        center:(CGPoint)origin
                      ringAxis:(TRAxis *)ringAxis
{
    int i;
    float radius, width;
    
    UIBezierPath *aPath = [[UIBezierPath alloc] init];
    
    // paint each even grid backgrounds
    if (_evenRowBackgroundColor != nil)
    {
        for (i = 0; i < ringAxis.numberOfTicks - 1; i++)
        {
            if (i % 2 == 0)
            {
                // radius of this ring
                radius = ([ringAxis position:(i + 1)] + [ringAxis position:i]) / 2.0;
                // width of this ring
                width = ([ringAxis position:(i + 1)] - [ringAxis position:i]);
                
                // add arc
                [aPath addArcWithCenter:origin radius:radius startAngle:-DEGREES_TO_RADIANS(size.right) endAngle:-DEGREES_TO_RADIANS(size.left) clockwise:FALSE];
                
                // draw the path
                CGContextBeginPath(ctx);
                CGContextAddPath(ctx, aPath.CGPath);
                
                CGContextSetLineWidth(ctx, width);
                CGContextSetStrokeColorWithColor(ctx, _evenRowBackgroundColor.CGColor);
                
                CGContextStrokePath(ctx);
                
                // clean up all path points
                [aPath removeAllPoints];
                
            }
        }
    }

    // paint each odd grid background
    if (_oddRowBackgroundColor != nil)
    {
        for (i = 0; i < ringAxis.numberOfTicks - 1; i++)
        {
            if (i % 2 != 0)
            {
                // radius of this ring
                radius = ([ringAxis position:(i + 1)] + [ringAxis position:i]) / 2.0;
                // width of this ring
                width = ([ringAxis position:(i + 1)] - [ringAxis position:i]);
                
                // add arc
                [aPath addArcWithCenter:origin radius:radius startAngle:-DEGREES_TO_RADIANS(size.right) endAngle:-DEGREES_TO_RADIANS(size.left) clockwise:FALSE];
                
                // draw the path
                CGContextBeginPath(ctx);
                CGContextAddPath(ctx, aPath.CGPath);
            
                CGContextSetLineWidth(ctx, width);
                CGContextSetStrokeColorWithColor(ctx, _oddRowBackgroundColor.CGColor);
                
                CGContextStrokePath(ctx);
                
                // clean up all path points
                [aPath removeAllPoints];
                
            }
        }
    }
}

// paint grid lines
-(void)paintGridLines:(CGContextRef)ctx
               layout:(Slice)size
               center:(CGPoint)origin
             ringAxis:(TRAxis *)ringAxis
            angleAxis:(TRAxis *)angleAxis

{
    // paint row grid lines
    [self paintRowGridLines:ctx
                     layout:size
                     center:origin
                   ringAxis:ringAxis
                  angleAxis:angleAxis];
    
    // paint column grid lines
    [self paintColumnGridLines:ctx
                        layout:size
                        center:origin
                      ringAxis:ringAxis
                     angleAxis:angleAxis];
    
    // paint focusing row or column
    if (_rowFocus >= 0 && _rowFocus < ringAxis.numberOfTicks)
    {
        [self paintRowGridLines_Neon:ctx
                              center:origin
                            ringAxis:ringAxis
                            rowFocus:_rowFocus
                          beginAngle:size.right
                            endAngle:size.left];
    }
    
    if (_columnFocus >= 0 && _columnFocus < angleAxis.numberOfTicks)
    {
        [self paintColumnGridLines_Neon:ctx
                                 center:origin
                              angleAxis:angleAxis
                            columnFocus:_columnFocus
                            beginRadius:size.bottom
                              endRadius:size.top];
    }
    
    return ;
}

// paint row grid line (x axis)
-(void)paintRowGridLines:(CGContextRef)ctx
                  layout:(Slice)size
                  center:(CGPoint)origin
                ringAxis:(TRAxis *)ringAxis
               angleAxis:(TRAxis *)angleAxis
{
    int i;
    float angle;
    UIBezierPath *aPath = [[UIBezierPath alloc] init];
    CGPoint bPt;
    
    // (2) draw ring (row) grid start from size.bottom to size.top
    // clean the path first
    float radius;
    [aPath removeAllPoints];
    
    angle = size.right;
    
    for (i = 0; i < ringAxis.numberOfTicks; i++)
    {
        // radius of this ring
        radius = [ringAxis position:i];
        
        // move to the begining angle line
        bPt.x = origin.x + radius * cosf(DEGREES_TO_RADIANS(angle));
        bPt.y = origin.y - radius * sinf(DEGREES_TO_RADIANS(angle));
        
        [aPath moveToPoint:bPt];
        
        // add arc
        [aPath addArcWithCenter:origin radius:radius startAngle:-DEGREES_TO_RADIANS(size.right) endAngle:-DEGREES_TO_RADIANS(size.left) clockwise:FALSE];
    }
    
    if (_rowGridLine.style == GridLine_Dash)
    {
        CGFloat dash[] = {2.0, 2.0};
        CGContextSetLineDash(ctx, .0, dash, 2);
    }
    
    // draw the path
    CGContextBeginPath(ctx);
    CGContextAddPath(ctx, aPath.CGPath);
    
    CGContextSetLineWidth(ctx, _rowGridLine.size);
    CGContextSetStrokeColorWithColor(ctx, _rowGridLine.color.CGColor);
    CGContextStrokePath(ctx);
    
    // Remove dash
    CGContextSetLineDash(ctx, .0, NULL, 0);
}

// paint column grid line (y axis)
-(void)paintColumnGridLines:(CGContextRef)ctx
                     layout:(Slice)size
                     center:(CGPoint)origin
                   ringAxis:(TRAxis *)ringAxis
                  angleAxis:(TRAxis *)angleAxis
{
    int i;
    float angle;
    CGPoint bPt, ePt;
    
    UIBezierPath *aPath = [[UIBezierPath alloc] init];
    
    // (1) drawing angle (column) grid start from the right to left
    // build path first
    for (i = 0; i < angleAxis.numberOfTicks; i++)
    {
        angle = [angleAxis position:i];
        
        // begin and end of angle grid line for angle
        bPt.x = origin.x + size.bottom * cosf(DEGREES_TO_RADIANS(angle));
        bPt.y = origin.y - size.bottom * sinf(DEGREES_TO_RADIANS(angle));
        
        ePt.x = origin.x + size.top * cosf(DEGREES_TO_RADIANS(angle));
        ePt.y = origin.y - size.top * sinf(DEGREES_TO_RADIANS(angle));
        
        [aPath moveToPoint:bPt];
        [aPath addLineToPoint:ePt];
    }
    
    if (_columnGridLine.style == GridLine_Dash)
    {
        CGFloat dash[] = {2.0, 2.0};
        CGContextSetLineDash(ctx, .0, dash, 2);
    }
    
    // draw the path
    CGContextBeginPath(ctx);
    CGContextAddPath(ctx, aPath.CGPath);
    
    CGContextSetLineWidth(ctx, _columnGridLine.size);
    CGContextSetStrokeColorWithColor(ctx, _columnGridLine.color.CGColor);
    CGContextStrokePath(ctx);
}

// paint neon light style type of line
-(void)paintRowGridLines_Neon:(CGContextRef)ctx
                       center:(CGPoint)origin
                     ringAxis:(TRAxis *)ringAxis
                     rowFocus:(int)rowFocus
                   beginAngle:(float)beginAngle
                     endAngle:(float)endAngle
{
    UIBezierPath *aPath = [[UIBezierPath alloc] init];
    CGPoint bPt;
    
    if (rowFocus >= ringAxis.numberOfTicks)
    {
        // do nothing
        return ;
    }
    
    float radius = [ringAxis position:rowFocus];
    
    // draw line around center in the {begin, end} range
    
    // move to the begining angle line
    bPt.x = origin.x + radius * cosf(DEGREES_TO_RADIANS(beginAngle));
    bPt.y = origin.y - radius * sinf(DEGREES_TO_RADIANS(beginAngle));
    
    [aPath moveToPoint:bPt];
    
    // add arc
    [aPath addArcWithCenter:origin radius:radius startAngle:-DEGREES_TO_RADIANS(beginAngle) endAngle:-DEGREES_TO_RADIANS(endAngle) clockwise:FALSE];
    
    // draw the path
    [self drawNeonPath:ctx
                 aPath:aPath
             lineColor:[UIColor blueColor].CGColor
             neonColor:[UIColor whiteColor].CGColor];
}

-(void)paintColumnGridLines_Neon:(CGContextRef)ctx
                          center:(CGPoint)origin
                       angleAxis:(TRAxis *)angleAxis
                     columnFocus:(int)columnFocus
                     beginRadius:(float)beginRadius
                       endRadius:(float)endRadius
{
    float angle;
    CGPoint bPt, ePt;
    
    UIBezierPath *aPath = [[UIBezierPath alloc] init];
    
    // (1) drawing angle (column) grid start from the right to left
    // build path first
    angle = [angleAxis position:columnFocus];
    
    // begin and end of angle grid line for angle
    bPt.x = origin.x + beginRadius * cosf(DEGREES_TO_RADIANS(angle));
    bPt.y = origin.y - beginRadius * sinf(DEGREES_TO_RADIANS(angle));
        
    ePt.x = origin.x + endRadius * cosf(DEGREES_TO_RADIANS(angle));
    ePt.y = origin.y - endRadius * sinf(DEGREES_TO_RADIANS(angle));
        
    [aPath moveToPoint:bPt];
    [aPath addLineToPoint:ePt];
    
    // draw the path
    [self drawNeonPath:ctx
                 aPath:aPath
             lineColor:[UIColor blueColor].CGColor
             neonColor:[UIColor whiteColor].CGColor];
}

// draw neon path
-(void)drawNeonPath:(CGContextRef)ctx
              aPath:(UIBezierPath *)aPath
          lineColor:(CGColorRef)lineColor
          neonColor:(CGColorRef)neonColor
{
    if (aPath == nil)
    {
        return ;
    }
    
    // draw the path
    CGContextBeginPath(ctx);
    
    CGContextAddPath(ctx, aPath.CGPath);
    // set outside shadow path
    CGContextSetShadowWithColor(ctx, CGSizeMake(0, 0), 20, neonColor);
    CGContextSetLineWidth(ctx, 5);
    CGContextSetStrokeColorWithColor(ctx, lineColor);
    CGContextStrokePath(ctx);
    
    CGContextAddPath(ctx, aPath.CGPath);
    CGContextSetShadowWithColor(ctx, CGSizeMake(0, 0), 10, lineColor);
    CGContextSetLineWidth(ctx, 3);
    CGContextSetStrokeColorWithColor(ctx, neonColor);
    CGContextStrokePath(ctx);
}

// paint grid table cells
-(void)paintCells:(CGContextRef)ctx
            table:(TRDataTableSliceView*)table
           layout:(Slice)size
           center:(CGPoint)origin
         ringAxis:(TRAxis *)ringAxis
        angleAxis:(TRAxis *)angleAxis
{
    // going over all cells of grid and display them on the map
    // row and column index
    int r, c;
    float angle, radius;
    // cell label point position
    CGPoint pt;
    
    if ([table numberOfColumns] != (angleAxis.numberOfTicks) ||
        [table numberOfRows] != ringAxis.numberOfTicks)
    {
        return ;
    }
    
    // all columns
    for (c = 0; c < angleAxis.numberOfTicks; c++)
    {
        angle = [angleAxis position:c];
        
        // all row of each column
        for (r = 0; r < ringAxis.numberOfTicks; r++)
        {
            // get cell object
            TRCell *cell = [table cell:r nCol:c];    
            if (cell == nil)
            {
                continue;
            }
            
            // compute grid view position for this cell
            radius = [ringAxis position:r];
            
            pt.x = origin.x + radius * cosf(DEGREES_TO_RADIANS(angle));
            pt.y = origin.y - radius * sinf(DEGREES_TO_RADIANS(angle));
            
            // show cell at position pt
            [self paintCell:ctx cell:cell position:pt];
        }
    }
}

// paint grid cell at given location
-(void)paintCell:(CGContextRef)ctx
            cell:(TRCell *)cell
        position:(CGPoint)position
{
    // check if we need to dynamicly draw cell image
    if (cell.image == nil)
    {
        // need to draw based on the properties
        return [self drawCell:ctx cell:cell position:position];
    }
    
    // we have preload image.  paint cell image at given location
    CGRect rect = CGRectMake(position.x - .5 * cell.image.size.width, position.y - .5 * cell.image.size.height, cell.image.size.width, cell.image.size.height);
    
    CGContextDrawImage(ctx, rect, cell.image.CGImage);
    
    return;
}

// dynamicly draw the cell at given location
-(void)drawCell:(CGContextRef)ctx
            cell:(TRCell *)cell
        position:(CGPoint)position
{
    // get cell drawing path
    UIBezierPath *cellPath = [self getCellPath:cell position:position];
    
    CGContextSaveGState(ctx);
    
    // (1) fill with subtle shadow
    if (cell.withShadow == TRUE)
    {
        CGContextSetShadowWithColor(ctx, CGSizeMake(0, 1), 1.0, [UIColor grayColor].CGColor);
    }
    
    CGContextSetFillColorWithColor(ctx, cell.metricColor.CGColor);
    
    CGContextAddPath(ctx, cellPath.CGPath);
    CGContextFillPath(ctx);

    // (2) outline
    if (cell.withOutline == TRUE)
    {
        CGContextSetStrokeColorWithColor(ctx, [UIColor grayColor].CGColor);
        CGContextSetLineWidth(ctx, .5);
        CGContextAddPath(ctx, cellPath.CGPath);
        CGContextStrokePath(ctx);
    }
    
    CGContextRestoreGState(ctx);
}

// get cell path
-(UIBezierPath *)getCellPath:(TRCell *)cell
                    position:(CGPoint)position
{
    float radius = cell.displaySize;
    
    if (cell.cellShape == CellValueShape_Circle || cell.cellShape == CellValueShape_Dot)
    {
        CGRect frame = CGRectMake(position.x - radius, position.y - radius, 2 * radius, 2 * radius);
        UIBezierPath *circle = [UIBezierPath bezierPathWithRoundedRect:frame
                                                      cornerRadius:frame.size.height * 1.0 / 2.0];
    
        return circle;
    }
    
    return nil;
}

@end

File: BackUpForABook/UIControls/TRClasses_swift/TRSliceGrid.swift
//
//  TRSliceGrid.swift
//  TreeRingMap_Swift
//
//  Created by dingf on 17/3/16.
//  Copyright © 2017年 MH.dingf. All rights reserved.
//

import UIKit

class TRSliceGrid: NSObject {
    var cellShowAsLayer: Bool?
    var columnGridLine: TRGridLine?
    var rowGridLine: TRGridLine?
    var rowFocus: Int?
    var columnFocus: Int?
    // even
    var evenRowBackgroundColor: UIColor?
    // odd
    var oddRowBackgroundColor: UIColor?
    
    var highlightedCellBackgroundColor: UIColor?
    var highlightedCellBorderColor: UIColor?
    
    
    override init() {
        cellShowAsLayer = true
        columnGridLine = TRGridLine.init()
        rowGridLine = TRGridLine.init()
        rowFocus = -1
        columnFocus = -1
        
    }
    
    func paint(_ ctx: CGContext, table: TRDataTableSliceView, size: Slice,origin: CGPoint, ringAxis: TRAxis, angleAxis: TRAxis){
        // paint grid background
        ctx.saveGState()
        paintGridRingBackground(ctx, size: size, origin: origin, ringAxis: ringAxis)
        ctx.restoreGState()
        // paint grid lines
        ctx.saveGState()
        paintGridLines(ctx, size: size, origin: origin, ringAxis: ringAxis, angleAxis: angleAxis)
        ctx.restoreGState()
        // paint grid cells
        if cellShowAsLayer == false{
            ctx.saveGState()
            paintCells(ctx, table: table, size: size, origin: origin, ringAxis: ringAxis, angleAxis: angleAxis)
        
        }
    }
    // paint grid background
    func paintGridRingBackground(_ ctx: CGContext, size: Slice, origin : CGPoint, ringAxis: TRAxis){
        let aPath = UIBezierPath()
        if evenRowBackgroundColor != nil && ringAxis.numberOfTicks! > 0{
            for i in 0..<ringAxis.numberOfTicks! - 1{
                if i % 2 == 0{
                    // radius of this ring
                    let radius = (ringAxis.position(i + 1) + ringAxis.position(i)) / 2.0
                    // width of this ring
                    let width = ringAxis.position(i + 1) - ringAxis.position(i)
                    // add arc
                    aPath.addArc(withCenter: origin, radius: CGFloat(radius), startAngle: -(CGFloat)(DEGREETORADIANS(degree: size.right!)), endAngle: -(CGFloat)(DEGREETORADIANS(degree: size.left!)), clockwise: false)
                    ctx.beginPath()
                    ctx.addPath(aPath.cgPath)
                    ctx.setLineWidth(CGFloat(width))
                    ctx.setStrokeColor(evenRowBackgroundColor!.cgColor)
                    ctx.strokePath()
                    aPath.removeAllPoints()
                }
            }
        }
        if oddRowBackgroundColor != nil && ringAxis.numberOfTicks! > 0{
            for i in 0..<ringAxis.numberOfTicks! - 1{
                if i % 2 != 0{
                    // radius of this ring
                    let radius = (ringAxis.position(i + 1) + ringAxis.position(i)) / 2.0
                    // width of this ring
                    let width = ringAxis.position(i + 1) - ringAxis.position(i)
                    // add arc
                    aPath.addArc(withCenter: origin, radius: CGFloat(radius), startAngle: -(CGFloat)(DEGREETORADIANS(degree: size.right!)), endAngle: -(CGFloat)(DEGREETORADIANS(degree: size.left!)), clockwise: false)
                    ctx.beginPath()
                    ctx.addPath(aPath.cgPath)
                    ctx.setLineWidth(CGFloat(width))
                    ctx.setStrokeColor(oddRowBackgroundColor!.cgColor)
                    ctx.strokePath()
                    aPath.removeAllPoints()
                }
            }
        }
    }
    
    // paint grid lines
    func paintGridLines(_ ctx: CGContext, size: Slice, origin: CGPoint, ringAxis: TRAxis, angleAxis: TRAxis){
        // paint row grid lines : circle ring
        paintRowGridLines(ctx, size: size, origin: origin, ringAxis: ringAxis, angleAxis: angleAxis)
        // paint column grid lines : line
        paintColumnGridLines(ctx, size: size, origin: origin, ringAxis: ringAxis, angleAxis: angleAxis)
        // paint focusing row or column
        if rowFocus! >= 0 && rowFocus! < ringAxis.numberOfTicks!{
            paintRowGridLines_Neon(ctx, origin: origin, ringAxis: ringAxis, rowFocus: rowFocus!, beginAngle: size.right!, endAngle: size.left!)
        }
        if columnFocus! >= 0 && columnFocus! < angleAxis.numberOfTicks!{
            paintColumnGridLine_Neon(ctx, origin: origin, angleAxis: angleAxis, columnFocus: columnFocus!, beginRadius: size.bottom!, endRadius: size.top!)
        }
        
    }
    
    // paint row grid line(x axis)
    func paintRowGridLines(_ ctx: CGContext, size: Slice, origin: CGPoint, ringAxis: TRAxis, angleAxis: TRAxis){
        let aPath = UIBezierPath()
        
        // draw row grid start from size.bottom to size.top
        let angle = size.right
        
        for i in 0..<ringAxis.numberOfTicks!{
            // radius fo this ring
            let radius = ringAxis.position(i)
            // move to the beginning angle line 
            let bpt = CGPoint(x: origin.x + CGFloat(radius * cosf(DEGREETORADIANS(degree: angle!))), y: origin.y - CGFloat(radius * sinf(DEGREETORADIANS(degree: angle!))))
            
            aPath.move(to: bpt)
            
            aPath.addArc(withCenter: origin, radius: CGFloat(radius), startAngle: -CGFloat(DEGREETORADIANS(degree: size.right!)), endAngle: -CGFloat(DEGREETORADIANS(degree: size.left!)), clockwise: false)
        }
        if rowGridLine?.style == .dash{
            let dash: [CGFloat] = [2.0, 2.0]
            ctx.setLineDash(phase: 0.0, lengths: dash)
        }
        // draw the path
        ctx.beginPath()
        ctx.addPath(aPath.cgPath)
        ctx.setLineWidth(CGFloat(rowGridLine!.size!))
        ctx.setStrokeColor(rowGridLine!.color!.cgColor)
        ctx.strokePath()
    }
    
    // paint column grid line(y axis)
    func paintColumnGridLines(_ ctx: CGContext, size: Slice, origin: CGPoint, ringAxis: TRAxis, angleAxis: TRAxis){
        let aPath = UIBezierPath()
        
        // draw column grid start from the right to left
        for i in 0..<angleAxis.numberOfTicks!{
            // radius fo this ring
            let angle = angleAxis.position(i)
            // move to the beginning angle line
            let bpt = CGPoint(x: origin.x + CGFloat(size.bottom! * cosf(DEGREETORADIANS(degree: angle))), y: origin.y - CGFloat(size.bottom! * sinf(DEGREETORADIANS(degree: angle))))
            
            let ept = CGPoint(x: origin.x + CGFloat(size.top! * cosf(DEGREETORADIANS(degree: angle))), y: origin.y - CGFloat(size.top! * sinf(DEGREETORADIANS(degree: angle))))
            aPath.move(to: bpt)
            aPath.addLine(to: ept)
        }
        if rowGridLine?.style == .dash{
            let dash: [CGFloat] = [2.0, 2.0]
            ctx.setLineDash(phase: 0.0, lengths: dash)
        }
        // draw the path
        ctx.beginPath()
        ctx.addPath(aPath.cgPath)
        ctx.setLineWidth(CGFloat(columnGridLine!.size!))
        ctx.setStrokeColor(columnGridLine!.color!.cgColor)
        ctx.strokePath()
      
    }
    
    // paint neon light style type of line 
    func paintRowGridLines_Neon(_ ctx: CGContext, origin: CGPoint, ringAxis: TRAxis, rowFocus: Int, beginAngle: Float, endAngle: Float){
        let aPath = UIBezierPath()
        if rowFocus >= ringAxis.numberOfTicks!{return}
        let radius = ringAxis.position(rowFocus)
        let bpt = CGPoint(x: origin.x + CGFloat(radius * cosf(DEGREETORADIANS(degree: beginAngle))), y: origin.y - CGFloat(radius * sinf(DEGREETORADIANS(degree: beginAngle))))
        aPath.move(to: bpt)
        aPath.addArc(withCenter: origin, radius: CGFloat(radius), startAngle: -(CGFloat)(DEGREETORADIANS(degree: beginAngle)), endAngle: -(CGFloat)(DEGREETORADIANS(degree: endAngle)), clockwise: false)
        
        drawNeonPath(ctx, aPath: aPath, lineColor: UIColor.blue.cgColor, neonColor: UIColor.white.cgColor)
    }
    
    func paintColumnGridLine_Neon(_ ctx: CGContext, origin: CGPoint, angleAxis: TRAxis, columnFocus: Int, beginRadius: Float, endRadius: Float){
        let aPath = UIBezierPath()
 
        let angle = angleAxis.position(columnFocus)
        let bpt = CGPoint(x: origin.x + CGFloat(beginRadius * cosf(DEGREETORADIANS(degree: angle))), y: origin.y - CGFloat(beginRadius * sinf(DEGREETORADIANS(degree: angle))))
        let ept = CGPoint(x: origin.x + CGFloat(endRadius * cosf(DEGREETORADIANS(degree: angle))), y: origin.y - CGFloat(endRadius * sinf(DEGREETORADIANS(degree: angle))))
        aPath.move(to: bpt)
        aPath.addLine(to: ept)
        
        drawNeonPath(ctx, aPath: aPath, lineColor: UIColor.blue.cgColor, neonColor: UIColor.white.cgColor)
    }
    
    // draw neon path
    func drawNeonPath(_ ctx: CGContext, aPath: UIBezierPath, lineColor: CGColor, neonColor: CGColor){
        ctx.setLineDash(phase: 0.0, lengths: [2.0, 0])
        ctx.beginPath()
        ctx.addPath(aPath.cgPath)
        ctx.setShadow(offset: CGSize.init(width: 0, height: 0), blur: 20.0, color: neonColor)
        ctx.setLineWidth(5)
        ctx.setStrokeColor(lineColor)
        ctx.strokePath()
        
        ctx.addPath(aPath.cgPath)
        ctx.setShadow(offset: CGSize.init(width: 0, height: 0), blur: 10.0, color: lineColor)
        ctx.setLineWidth(3.0)
        ctx.setStrokeColor(neonColor)
        ctx.strokePath()
    }
    
    // paint grid table cells
    func paintCells(_ ctx: CGContext, table: TRDataTableSliceView, size: Slice, origin: CGPoint, ringAxis: TRAxis, angleAxis: TRAxis){
        for c in 0..<angleAxis.numberOfTicks!{
            let angle = angleAxis.position(c)
            // all row of each column
            for r in 0..<ringAxis.numberOfTicks!{
                // get cell object
                let cell = table.cellOfRowAndCol(r, col: c)
                let radius = ringAxis.position(r)
                let pt = CGPoint(x: origin.x + CGFloat(radius * cosf(DEGREETORADIANS(degree: angle))), y: origin.y - CGFloat(radius * sinf(DEGREETORADIANS(degree: angle))))
                // show cell at position pt
                if cell.image == nil {
                    let cellPath = getCellPath(cell, position: pt)
                    ctx.saveGState()
                    // fill with subtle shadow
                    if cell.withShadow == true{
                        ctx.setShadow(offset: CGSize.init(width: 0, height: 1), blur: 1.0, color: UIColor.gray.cgColor)
                    }
                    ctx.setFillColor(cell.metricColor!.cgColor)
                    ctx.addPath(cellPath!.cgPath)
                    ctx.fillPath()
                    
                    // outline
                    if cell.withOutline == true{
                        ctx.setStrokeColor(UIColor.gray.cgColor)
                        ctx.setLineWidth(0.5)
                        ctx.addPath(cellPath!.cgPath)
                        ctx.strokePath()
                    }
                    ctx.restoreGState()
                }
                // else wo have preload image
                let rect = CGRect(x: pt.x - 0.5 * cell.image!.size.width, y: pt.y - 0.5 * cell.image!.size.height, width: cell.image!.size.width, height: cell.image!.size.height)
                ctx.draw(cell.image!.cgImage!, in: rect)
            }
        }
    }
    
    // get cell path
    func getCellPath(_ cell: TRCell, position: CGPoint) -> UIBezierPath?{
        let radius = cell.displaySize
        if cell.cellShape == .circle || cell.cellShape == .dot{
            let frame = CGRect(x: position.x - CGFloat(radius!), y: position.y - CGFloat(radius!), width: 2 * CGFloat(radius!), height: 2 * CGFloat(radius!))
            let circle = UIBezierPath(roundedRect: frame, cornerRadius: frame.size.height * 1.0 / 2.0 )
            return circle
        }
        return nil
    }

}
